docs/design/horizontal-pod-autoscaler.md

```diff
@@ -0,0 +1,269 @@
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Horizontal Pod Autoscaling
+
+## Preface
+
+This document briefly describes the design of the horizontal autoscaler for pods.
+The autoscaler (implemented as a kubernetes control loop) is responsible for automatically
+choosing and setting the number of pods of a given type that run in a kubernetes cluster.
+
+This design supersedes [autoscaling.md](http://releases.k8s.io/release-1.0/docs/proposals/autoscaling.md).
+
+## Overview
+
+The usage of a serving application usually vary over time: sometimes the demand for the application rises,
+and sometimes it drops.
+In Kubernetes version 1.0, a user can only manually set the number of serving pods.
+Our aim is to provide a mechanism for the automatic adjustment of the number of pods based on CPU utilization statistics.
+
+## Scale Subresource
+
+In Kubernetes version 1.1, we are introducing Scale subresource and implementing horizontal autoscaling of pods based on it.
+Scale subresource is supported for replication controllers and deployments.
+Scale subresource is a Virtual Resource (is not stored in etcd as a separate object).
+It is only present in API as an interface to accessing replication controller or deployment,
+and the values of Scale fields are inferred from the corresponding replication controller/deployment object.
+HorizontalPodAutoscaler object is bound with exactly one Scale subresource and is
+autoscaling associated replication controller/deployment through it.
+The main advantage of such approach is that whenever we introduce another type we want to auto-scale,
+we just need to implement Scale subresource for it (w/o modifying autoscaler code or API).
+The wider discussion regarding Scale took place in [#1629](https://github.com/kubernetes/kubernetes/issues/1629).
+
+Scale subresource is in API for replication controller or deployment under the following paths:
+
+```apis/extensions/v1beta1/replicationcontrollers/myrc/scale```
+
+```apis/extensions/v1beta1/deployments/mydeployment/scale```
+
+It has the following structure:
+
+```go
+// represents a scaling request for a resource.
+type Scale struct {
+	unversioned.TypeMeta
+	api.ObjectMeta
+
+	// defines the behavior of the scale.
+	Spec ScaleSpec
+
+	// current status of the scale.
+	Status ScaleStatus
+}
+
+// describes the attributes of a scale subresource
+type ScaleSpec struct {
+	// desired number of instances for the scaled object.
+	Replicas int `json:"replicas,omitempty"`
+}
+
+// represents the current status of a scale subresource.
+type ScaleStatus struct {
+	// actual number of observed instances of the scaled object.
+	Replicas int `json:"replicas"`
+
+	// label query over pods that should match the replicas count.
+	Selector map[string]string `json:"selector,omitempty"`
+}
+```
+
+Writing ```ScaleSpec.Replicas``` resizes the replication controller/deployment associated with
```

Comment:
Don't use three backticks for inline quoting -- it screws up the Github diff view coloring. Use one backtick when it's inline, three when it's on a new line.

Also, s/Writing/Writing to/